### PR TITLE
Include digest of ProtocolConfig when voting for a version

### PR DESCRIPTION
**Pull request by mystenmark**
Add AuthorityCapabilitiesV2, which includes the digest of the protocol config for each version.

This makes it impossible for the committee to approve a protocol upgrade if (due to a release mistake) there is disagreement about what is in the config.

### CODE REVIEW

File: crates/sui-core/src/authority.rs [hunk 1]
```diff
@@ -4374,7 +4374,95 @@ impl AuthorityState {
             })
     }
 
-    fn choose_protocol_version_and_system_packages(
+    fn is_protocol_version_supported_v2(
```
Comment (lxfind) [inline crates/sui-core/src/authority.rs:4377]: I understand v1 and v2 uses different internal data structures, but do you think it's possible to find a way to dedup some code?

Comment (mystenmark) [reply crates/sui-core/src/authority.rs:4377]: I did originally have a significantly dedupped version that was based on translated v1 to v2 when loading it from the database. But:

1. It was hard to feel confident that that code behaved 100% identically, and therefore couldn't cause a fork
2. v1 can be deleted once v2 is rolled out

File: crates/sui-types/src/supported_protocol_versions.rs [hunk 2]
```diff
@@ -36,4 +41,61 @@ impl SupportedProtocolVersions {
...
+    // Ideally this would be in sui-protocol-config, but sui-types depends on sui-protocol-config,
+    // so it would introduce a circular dependency.
+    fn protocol_config_digest(config: &ProtocolConfig) -> Digest {
+        let intent = Intent {
+            scope: IntentScope::ProtocolConfigDigest,
+            version: IntentVersion::V0,
+            app_id: AppId::Sui,
+        };
+        let mut digest = DefaultHash::default();
+        bcs::serialize_into(&mut digest, &intent).expect("serialization cannot fail");
```
Comment (lxfind) [inline crates/sui-types/src/supported_protocol_versions.rs:80]: Could you explain why we need to include the intent in the digest computation? (and comments too) It doesn't seem necessary to me

Comment (mystenmark) [reply crates/sui-types/src/supported_protocol_versions.rs:80]: I suppose we don't, there's no need for domain separation here. Was just copying the typical pattern from elsewhere in the code.

**Review by lxfind — APPROVED** [review]
LGTM

File: crates/sui-core/src/authority.rs [hunk 3]
```diff
@@ -4287,7 +4287,7 @@ impl AuthorityState {
         Some(res)
     }
 
-    fn is_protocol_version_supported(
+    fn is_protocol_version_supported_v1(
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/authority.rs:4320]: done

**Pull request merged**